command/alloc_exec.go

```diff
@@ -0,0 +1,338 @@
+		return 1
+	}
+	args = flags.Args()
+
+	// try to infer stdin value if one option doesn't match defaults
+	stdinOpt := stdinShortOpt && stdinLongOpt
+
+	ttyOpt := stdinTty
+	if ttyLongOpt != stdinTty || ttyShortOpt != stdinTty {
+		ttyOpt = !stdinTty
+	}
+
+	if !stdinOpt {
+		// if stdin is disabled, disable tty
+		// TODO: detect if user passed incompatible settings and report
+		ttyOpt = false
+	}
+
+	if numArgs := len(args); numArgs < 1 {
+		if job {
+			l.Ui.Error("A job ID is required")
+		} else {
+			l.Ui.Error("An allocation ID is required")
+		}
+
+		l.Ui.Error(commandErrorText(l))
+		return 1
+	} else if numArgs < 2 {
+		l.Ui.Error("A command is required")
+		l.Ui.Error(commandErrorText(l))
+		return 1
+	}
+
+	command := args[1:]
+
+	client, err := l.Meta.Client()
+	if err != nil {
+		l.Ui.Error(fmt.Sprintf("Error initializing client: %v", err))
+		return 1
+	}
+
+	// If -job is specified, use random allocation, otherwise use provided allocation
+	allocID := args[0]
+	if job {
+		allocID, err = getRandomJobAlloc(client, args[0])
+		if err != nil {
+			l.Ui.Error(fmt.Sprintf("Error fetching allocations: %v", err))
+			return 1
+		}
+	}
+
+	length := shortId
+
+	// Query the allocation info
+	if len(allocID) == 1 {
+		l.Ui.Error(fmt.Sprintf("Alloc ID must contain at least two characters."))
+		return 1
+	}
+
+	allocID = sanitizeUUIDPrefix(allocID)
+	allocs, _, err := client.Allocations().PrefixList(allocID)
+	if err != nil {
+		l.Ui.Error(fmt.Sprintf("Error querying allocation: %v", err))
+		return 1
+	}
+	if len(allocs) == 0 {
+		l.Ui.Error(fmt.Sprintf("No allocation(s) with prefix or id %q found", allocID))
+		return 1
+	}
+	if len(allocs) > 1 {
+		// Format the allocs
+		out := formatAllocListStubs(allocs, false, length)
+		l.Ui.Error(fmt.Sprintf("Prefix matched multiple allocations\n\n%s", out))
+		return 1
+	}
+	// Prefix lookup matched a single allocation
+	alloc, _, err := client.Allocations().Info(allocs[0].ID, nil)
+	if err != nil {
+		l.Ui.Error(fmt.Sprintf("Error querying allocation: %s", err))
+		return 1
+	}
+
+	if task == "" {
+		// Try to determine the tasks name from the allocation
+		var tasks []*api.Task
+		for _, tg := range alloc.Job.TaskGroups {
+			if *tg.Name == alloc.TaskGroup {
+				if len(tg.Tasks) == 1 {
+					task = tg.Tasks[0].Name
+					break
+				}
+
+				tasks = tg.Tasks
+				break
+			}
+		}
+
+		if task == "" {
+			l.Ui.Error(fmt.Sprintf("Allocation %q is running the following tasks:", limit(alloc.ID, length)))
+			for _, t := range tasks {
+				l.Ui.Error(fmt.Sprintf("  * %s", t.Name))
+			}
+			l.Ui.Error("\nPlease specify the task.")
+			return 1
+		}
+	}
+
+	if l.Stdin == nil {
+		l.Stdin = os.Stdin
+	}
+	if l.Stdout == nil {
+		l.Stdout = os.Stdout
+	}
+	if l.Stderr == nil {
+		l.Stderr = os.Stderr
+	}
+
+	var stdin io.Reader = l.Stdin
+	if !stdinOpt {
+		stdin = bytes.NewReader(nil)
+	}
+
+	code, err := l.execImpl(client, alloc, task, ttyOpt, command, stdin, l.Stdout, l.Stderr)
```

Comment:
you may want to validate the task exists in the allocation early? (`func validateTaskExistsInAllocation(taskName string, alloc *api.Allocation) error`) in `alloc_restart.go` handles this for you.